### DISCUSSION TITLE
Vite error (document variable not found)

### DISCUSSION THREAD

**Discussion by AndrewJSchoen** [remarkjs/remark#1083] [Q&A] [answered]
Hi all, 

I have a component library that is using `react-markdown`, and I am building it using Vite.js/Storybook. However, after building and using it in another project, I get an error that:

```
ReferenceError: Can't find variable: document
```

This seems to originate in some built code that looks like this:
```
const c4 = document.createElement("i");
function YP(e) {
  const t = "&" + e + ";";
  c4.innerHTML = t;
  const n = c4.textContent;
  return n.charCodeAt(n.length - 1) === 59 && e !== "semi" || n === t ? !1 : n;
}
```

This happens in even the most simple of examples:
```
<ReactMarkdown>
  # Hello World
</ReactMarkdown>
```

However, I am not sure if this is a Vite issue, or one specific to `react-markdown`. Does anyone know if there is a setting with `react-markdown` that I am missing that would make this work, or would it not work the way I am trying to use it? I should note that I am building using storybook, and it functions fine in that environment.

**Comment by ChristianMurphy** [comment, marked as answer]
Most likely it is an issue with a specific vite or storybook configuration option.
React Markdown works fine with vite, see the following example sandbox: https://stackblitz.com/edit/vitejs-vite-hxte3z?file=main.jsx

<details>
<summary>key files from sandbox</summary>

_vite.config.js_
```js
import { defineConfig } from 'vite';
import react from '@vitejs/plugin-react';

export default defineConfig({
  plugins: [react()],
});
```

_main.jsx_
```jsx
import ReactDom from 'react-dom';
import ReactMarkdown from 'react-markdown';

const domContainer = document.querySelector('#app');
const root = ReactDom.createRoot(domContainer);
root.render(<ReactMarkdown>{`# header`}</ReactMarkdown>);
```

_package.json_
```json
{
  "name": "vite-react-markdown-example",
  "private": true,
  "version": "0.0.0",
  "type": "module",
  "scripts": {
    "dev": "vite",
    "build": "vite build",
    "preview": "vite preview"
  },
  "devDependencies": {
    "@vitejs/plugin-react": "^2.0.0",
    "vite": "^4.0.0-beta.5"
  },
  "dependencies": {
    "react": "^18.0.0",
    "react-dom": "^18.0.0",
    "react-markdown": "^8.0.0"
  }
}
```

</details>

It is definitely not directly coming from `react-markdown`, it does not use `document.createElement` or `.innerHTML` anywhere it the code base.
https://github.com/remarkjs/react-markdown/search?q=innerHTML
https://github.com/remarkjs/react-markdown/search?q=document.createElement

**Reply by AndrewJSchoen** [reply]
Thanks for the examples and info. Yeah, it is pretty hard to diagnose, since as I mentioned, it works perfectly fine when _within_ the storybook environment, but when built and imported elsewhere it causes the error.

**Reply by ChristianMurphy** [reply]
It can be, https://github.com/remarkjs/.github/blob/main/support.md#asking-quality-questions offers some advise which can help narrow things down.
In particular a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) can be a huge help both for you better understanding the issue, and to share in a form that others can more meaningfully contribute to.
Right now, it's a bit of a shot in the dark guessing what could be going on.

Taking another random guess:
Are you using server side rendering?
This could be the reverse of https://github.com/vfile/vfile/discussions/65
Where vite may be incorrectly choosing the browser version, instead of node/server friendly version, for server side rendering.

**Reply by AndrewJSchoen** [reply]
Given that the error only appears when _loading_ the built library in another project, I'm not sure there is an easy way to create a single codesandbox that illustrates the problem (but if there is, I would be interested). That being said, I tested out the `react-remark` library, swapping out the `ReactMarkdown` component for the `Remark` component, and it seems like the error is not present. If I can customize how the components are rendered, [which it seems it is](https://github.com/remarkjs/react-remark#pass-options-to-component), then I should have _a_ solution.

**Reply by remcohaszing** [reply]
I have seen this problem before. It was because the code was running inside a worker. Could that be the cause?

The [workaround](https://github.com/remcohaszing/monaco-unified/blob/main/examples/demo/webpack.config.js#L18..L20) was to explicitly load the default package export of `decode-named-character-reference` instead of the browser export.

**Reply by AndrewJSchoen** [reply]
@remcohaszing Maybe? I was using my library (that used `ReactMarkdown`) in a project that used workers (specifically `comlink`), but they weren't explicitly dependent on one another. That being said, the reason the issue became problematic was because the web worker itself wasn't working anymore after adding `ReactMarkdown`. It turned out that a 1:1 swap between the `ReactMarkdown` component and the `Remark` component did end up fixing the issue with no real refactoring needed on my end, so I just did that.

**Reply by remcohaszing** [reply]
My guess is that it works, because `react-remark` uses an [outdated version](https://github.com/remarkjs/react-remark/blob/main/package.json#L41) of `remark-parse`.

**Reply by AndrewJSchoen** [reply]
@remcohaszing Just to clarify, since this may become an issue if that gets updated, should the workaround you specify be done in the library or the project using it?

**Reply by remcohaszing** [reply]
As can be seen [here](https://github.com/wooorm/decode-named-character-reference/blob/1.0.2/package.json#L32-L40), `decode-named-character-reference` specifies the `worker` and `browser` exports. The `browser` export uses the DOM. All other build targets use the `character-entities` dependency. That library is configured correctly. The issue is that your build tools should use the `worker` export for web workers.

You could also choose not to update `react-remark` or use an older version of `react-markdown`, which is effectively the solution you choose by using `react-remark` now.
